### ISSUE TITLE
`possibly-used-before-assigment` fails to understand test with immutable `bool`

### ISSUE THREAD

**Issue by lucascolley** [pylint-dev/pylint#10095]
### Bug description

As per the docs, this works:

```python
def func(x: bool) -> None:
    if not x:
        msg = 'hello world'
    print("hi")
    if not x:
        print(msg)
    return
```

but only because we are using the same test, `not x`, twice.

If we instead use the test `x`, pylint fails:

```python
def func(x: bool) -> None:
    if not x:
        msg = 'hello world'
    print("hi")
    if x:
        return
    print(msg)
```

### Command used

```shell
pylint
```

### Pylint output

```python
E0606: Possibly using variable 'msg' before assignment (possibly-used-before-assignment)
```

### Expected behavior

It is impossible for `print("hi")` (or any intervening code that doesn't assign to `x`) to change the value of `x` since it is an immutable bool. If `not x` is `False` at the top of the function, `x` will always be `True` at the bottom of the function. So it is impossible for `msg` to be used before assignment - this is a false positive.

### Pylint version

```shell
pylint 3.3.1
```

**Comment by jacobtylerwalls** [comment]
Thanks for the report, this is an even simpler case than the one discussed in #9662, for which I would expect any fix to also cover this case. 👍

**Comment by jacobtylerwalls** [comment]
> So it is impossible for msg to be used before assignment - this is a false positive.

Part of the premise that I want to make explicit here is that there are no other reassignments to `x` in the middle of the function. `x = not x` before the second test at the bottom of the function would violate your example.
